### ISSUE TITLE
customize the name of the .aiida folder at installation? (and managing multiple profiles)

### ISSUE THREAD

**Issue by nmounet** [aiidateam/aiida_core#404]
Even using virtual environments, one cannot easily switch between very different AiiDA versions (e.g. if there are configuration change requiring a migration), because the .aiida folder is always named in the same way.

**Comment by lekah** [comment]
Very important issue, I think. I am using a bash-script now to switch environment and daemon user every now and then, which is not a good way, I think. Can one have multiple brokers running, in princpiple? (That's the real question, I think, not whether you can customize folders)

**Comment by greschd** [comment]
I think it might work, as long as the two broker URLs are distinct, and the two celery apps don't clash. For the celery apps it could be enough that celery is installed in each virtualenv separately.

I'm running a setup with manually changed ``.aiida`` folder, and will report if I run into any troubles.

**Comment by giovannipizzi** [comment]
I use this issue for a more general discussion we have been having on how to manage multiple profiles.

One interesting option we were thinking is to keep the `~/.aiida` folder as a "global" settings folder, but allow any folder on disk to have a, .e.g, .`~/.aiidaprofile` folder which would contain another config.json file. If the user runs `verdi` from within one of these folders (or its subfolders), this config file overrides the global one. 

The idea is to map what git does (when it looks if you are in a git repo).

Pro: 
* it's much easier to work on different projects using different profiles, you just have to change folder

Con/to discuss:
* what would be the content of the folder? If it's just an empty folder maybe it does not make sense. However I typically have a number of scripts I use to run simulations and post-process the data of a given profile in a folder. We might suggest to people to have this into that folder? This could practically happen by having a command to create a barebone folder structure with a few folders beside the `~/.aiidaprofile`?
* we should be careful about storing credentials in the config.json, if people want to put the whole folder under e.g. git - maybe we can keep credentials in the global `~/.aiida` folder, and just move the information in `~/.aiidaprofile`? Or even better keep all profiles in `~/.aiida` as it is now, and put in `~/.aiidaprofile` only the information on what is the default profile to use if we are in the folder (and maybe which virtualenv, which profile for the daemon, ...)?
* we should check how fast we can do the check that git does (since git does it, my guess is that it can be done reasonably fast - I would say that we need it to run in <50ms). The logic should be: check if in this folder there is a `.aiidaprofile/config.json` file (and then use it), otherwise iterate on the folder above until you hit a filesystem boundary (not necessarily only `/`). Also, we should probably make sure that infinite loops are not possible e.g. with symlinks (but probably this is not possible when going up, only when going down in folders?)
* I think if we go along this route, we should have an option *not* to have a default AiiDA profile if you are outside any folder (to avoid confusion; the user would instead get an error message). I would make this the default, and of course if the user wants they could choose to have one.

Would this solve/help in your usecases above @nmounet @lekah @greschd ?

**Comment by greschd** [comment]
I think that's a very promising idea. A few thoughts:

* I think the content of the folder should be the same as the global ``.aiida`` -- essentially there shouldn't be a problem when the global ``~/.aiida`` is missing, as long as there's a local one. One advantage of this is that if there's ever a need to change the layout of ``.aiida`` between versions, you could still keep different versions "alive" at the same time.
* For the credentials, maybe we should just make them a separate file (something like ``.aiidapass``)? That way the config could contain all non-sensitive data, and people can just put the password file in their ``.gitignore``. 
* My workflow is usually to change virtualenv for different configurations, and not necessarily folder. Obviously I'll go to the folder at some point, but just opening up a terminal and switching to the right venv is much quicker (e.g. to use ``verdi shell``). So maybe we could add something like an ``AIIDA_PATH`` environment variable to configure where the search for the ``.aiidaprofile`` folder starts? It could just default to ``.`` to get back to the previous behavior. Or maybe some lookup rules like with ``PYTHONPATH`` are appropriate?
* Naive implementation of getting the folder that takes 125 µs (tested going up 7 directories). I'm sure you can get faster than that, but the point is that it's very cheap.

```python
import os
import itertools
try:
    from pathlib import Path
except ImportError:
    from pathlib2 import Path

def find_path(root, dir_name):
    path = Path(os.path.abspath(root))
    for p in itertools.chain([path], path.parents):
        directory = p / dir_name
        if directory.is_dir():
            return directory
    raise OSError('No directory found')
```

**Comment by greschd** [comment]
I've found another use case for this feature:

I tried to create a testing infrastructure (based on ``pytest`` and ``pgtest``) that sets up everything on an isolated temporary database, and a separate temporary directory for config + repository. To get AiiDA to use the right config I just monkeypatched ``AIIDA_CONFIG_FOLDER``.

This worked pretty well (setting up everything from within code is a bit tricky, but I think this will improve with the move to ``click``), but the problem is that I cannot launch the daemon pointing to the right DB / config file. 

The takeaway for this is that I'm strongly in favor of having an environment variable (``AIIDA_PATH``) which acts similar to ``PYTHONPATH`` in setting an order of precedence where to look for the ``.aiida`` (or ``.aiidaprofile``) folder. 
Something to think about is whether we want to prefer exact matches over those where the folder is higher up (if you have ``AIIDA_PATH='/foo/bar:/baz'``, and there are ``/foo/.aiida`` and ``/baz/.aiida``, which one do you prefer?).

**Comment by szoupanos** [comment]
A few comments/ideas on this discussion:

- It seems to me that there are two (orthogonal) issues: (a) Running (concurrently?) different AiiDA versions and (b) switching properly between different AiiDA profiles. I don’t believe that (a) should be used as a solution to (b). It might be easy in the beginning but chaotic (?) in the long run since it will need different AiiDA installations per profile.
- Switching between profiles should not impose stopping the AiiDA daemon. We can/should work on that, since it is not the case today. It should not be very difficult to achieve it. In general, running concurrently different profiles of a specific AiiDA version/installation (even with different backends) should be possible.
- AiiDA should be (?) responsible for all the profiles related to a specific version (i.e. it’s very likely that the config.json may change from one version to another).
- It is good a version of AiiDA to be self contained when running in a virtual environment environment (like pointing to the correct .aiida with a venv variable like Dominik mentioned). However, we should take care of not having daemon “collisions”. In general, AiiDA should be able to have daemons of different profiles running simultaneously and (as a second level) we should be able to run (simultaneously) different versions of AiiDA (with each of them having a number of daemons running).

Let me know what you also think or if I misunderstood something.

**Comment by lekah** [comment]
Hi all, thanks for the comments. I have some of my own:

- I think that the idea of recursing upwards until you find a matching folder creates more problems than it solves. For example, what if you have a folder of python scripts that you run routinely with different profiles? We now have an infrastructure setup that allows doing that very easily, so let's not get rid of that.
- The folder structure now is also quite flexible, since we have ~/.aiida/different-profiles/... Again, something that works well for us and is in my opinion worth preserving.

As @szoupanos  said, there are two issues that can be separated:

- Switching profiles: Above idea in a different way: the easiest solution would be to export an AIIDA_PROFILE variable whenever you switch environment. This is a very easy solution and can be implemented with 2 additional lines. I.e. the default profile is not read from the JSON but from OS environment. 
- Multiple daemons: The only problem now is how to run daemons for different profiles simultaneously. But this is not a big problem, we could simply copy the daemon folder into each profile's repository. This way each profile has a daemon and its own logfile and there's no mixing up things.

Would this solve the issues?  We could still have config-files in the individual repository, and the ~/.aiida/config.json contains only the mapping between profile and the repository path?

**Comment by greschd** [comment]
I think it's a good point that this doesn't really solve the "multiple daemons" problem. Maybe we should create a separate issue for this problem? If we decide that it's not worth doing we can always close it.

@lekah, I think you misunderstood how the "recursing up" works (well, at least in my implementation in #616). We go through the list of paths in ``AIIDA_PATH``, and do the recursing-upwards from each of these. So in fact you can just export ``AIIDA_PATH`` to be a specific path when switching virtualenvs. For my purpose, I also added the current folder (``AIIDA_PATH='.'``) so that when I switch to the tests folder I automatically get the test configuration. But I'd say the most 'common' way to use it would be to just set it to a per-virtualenv configuration folder.

**Comment by greschd** [comment]
@szoupanos I don't really agree that one AiiDA install should **always** be responsible for all profiles related to it. I have two use cases right now where that's not the case:
* Multiple users sharing the same AiiDA installation (on our "managed" AiiDA server)
* Test profiles which should not interact with the production config at all.

However, I agree that all profiles should be in the same place in **most** cases -- and that this is how we should present it in the documentation (as "default" setup).

**Comment by szoupanos** [comment]
@greschd 
One more question. 

How does the current testing infrastructure limit you? Or to say it differently, in which direction you would like to improve it?

Currently, the testing profiles don't interact with the non-testing profiles (they have their own database and repository). They are listed in the main .aiida/config.json but they don't interact with the user profiles. Does this behaviour limit you?

**Comment by greschd** [comment]
@szoupanos For my plugins I'm currently not using the built-in AiiDA testing infrastructure. Since there's no entrypoint for plugin tests (yet?), I decided to take a bit of a different approach:

When tests are run, a temporary postgres cluster and repository directory are created. The whole setup is then done automatically, where the computers / codes are specified via config files. Finally the daemon is started, and then the test cases run. I used pytest since that's what I'm most comfortable with.

Anyway, I'm not really sure if that's the right approach. The benefit is that you can have truly independent tests, where the setup of the DB / profile is formalized in code. On the other hand, there is some overhead to run the tests (a few seconds for DB creation and daemon start-stop each).

In principle it could probably work if the daemon can just run on different profiles simultaneously. The only complication is that since the temporary DB and repository are different in each run, the profile has to be re-configured accordingly. Also, it might be a bit ugly to have test profiles for each plugin. But then again, every plugin development should probably have a separate virtualenv / AiiDA install.

**Comment by greschd** [comment]
Closed by #616.
